js/src/admin/components/AdminPage.tsx

```diff
@@ -0,0 +1,290 @@
+import Page from '../../common/components/Page';
+import Button from '../../common/components/Button';
+import Switch from '../../common/components/Switch';
+import Select from '../../common/components/Select';
+import classList from '../../common/utils/classList';
+import Stream from '../../common/utils/Stream';
+import saveSettings from '../utils/saveSettings';
+import AdminHeader from './AdminHeader';
+import type Mithril from 'mithril';
+
+interface AdminHeaderOptions {
+  title: string;
+  description: string;
+  icon: string;
+  /**
+   * Will be used as the class for the AdminPage.
+   *
+   * Will also be appended with `-header` and set as the class for the `AdminHeader` component.
+   */
+  className: string;
+}
+
+type HTMLInputTypes =
+  | 'button'
+  | 'checkbox'
+  | 'color'
+  | 'date'
+  | 'datetime-local'
+  | 'email'
+  | 'file'
+  | 'hidden'
+  | 'image'
+  | 'month'
+  | 'number'
+  | 'password'
+  | 'radio'
+  | 'range'
+  | 'reset'
+  | 'search'
+  | 'submit'
+  | 'tel'
+  | 'text'
+  | 'time'
+  | 'url'
+  | 'week';
+
+interface CommonSettingsItemOptions extends Mithril.Attributes {
+  setting: string;
+  label: string | ReturnType<typeof app.translator.trans>;
+  help?: string | ReturnType<typeof app.translator.trans>;
+  className?: string;
+}
+
+/**
+ * Valid options for the setting component builder to generate an HTML input element.
+ */
+export interface HTMLInputSettingsComponentOptions extends CommonSettingsItemOptions {
+  /**
+   * Any valid HTML input `type` value.
+   */
+  type: HTMLInputTypes;
+}
+
+/**
+ * Valid options for the setting component builder to generate a Switch.
+ */
+export interface SwitchSettingComponentOptions extends CommonSettingsItemOptions {
+  type: 'bool' | 'checkbox' | 'switch' | 'boolean';
+}
+
+/**
+ * Valid options for the setting component builder to generate a Select dropdown.
+ */
+export interface SelectSettingComponentOptions extends CommonSettingsItemOptions {
+  type: 'select' | 'dropdown' | 'selectdropdown';
+  /**
+   * Map of values to their labels
+   */
+  options: { [value: string]: string | ReturnType<typeof app.translator.trans> };
+  default: string;
+}
+
+/**
+ * All valid options for the setting component builder.
+ */
+export type SettingsComponentOptions = HTMLInputSettingsComponentOptions | SwitchSettingComponentOptions | SelectSettingComponentOptions;
+
+/**
+ * Valid attrs that can be returned by the `headerInfo` function
+ */
+export type AdminHeaderAttrs = AdminHeaderOptions & Partial<Omit<Mithril.Attributes, 'class'>>;
+
+export default class AdminPage extends Page {
+  settings: Record<string, () => string> = {};
```

Comment:
Can we import the mithril `Stream` type instead and use that ? because `() => string` isn't accurate.

```ts
import type StreamType from 'mithril/stream';

export default class AdminPage extends Page {
  settings: Record<string, StreamType<string>> = {};
```

Reply:
Oops! Done! 👍